lib/internal/fs/cp/cp.js

```diff
@@ -0,0 +1,402 @@
+'use strict';
+
+// This file is a modified version of the fs-extra's copy method
+// modified in the following ways:
+//
+// - Use of the graceful-fs has been replaced with fs.
+// - Helpers from stat.js have been inlined.
+// - Formatting and style changes to match Node.js' linting rules.
+// - Parameter validation has been added.
+
+const {
+  ArrayPrototypeEvery,
+  ArrayPrototypeFilter,
+  Boolean,
+  PromiseAll,
+  PromisePrototypeCatch,
+  PromisePrototypeThen,
+  PromiseReject,
+  SafeArrayIterator,
+  StringPrototypeSplit,
+} = primordials;
+const {
+  codes: {
+    ERR_FS_CP_DIR_TO_NON_DIR,
+    ERR_FS_CP_EEXIST,
+    ERR_FS_CP_FIFO_PIPE,
+    ERR_FS_CP_NON_DIR_TO_DIR,
+    ERR_FS_CP_SOCKET,
+    ERR_FS_CP_SYMLINK_TO_SUBDIRECTORY,
+    ERR_FS_CP_TO_SUBDIRECTORY,
+    ERR_FS_CP_UNKNOWN,
+    ERR_FS_EISDIR,
+  },
+} = require('internal/errors');
+const {
+  os: {
+    errno: {
+      EEXIST,
+      EISDIR,
+      EINVAL,
+      ENOTDIR,
+    }
+  }
+} = internalBinding('constants');
+const {
+  chmod,
+  copyFile,
+  lstat,
+  mkdir,
+  readdir,
+  readlink,
+  stat,
+  symlink,
+  unlink,
+  utimes,
+} = require('fs/promises');
+const {
+  dirname,
+  isAbsolute,
+  join,
+  parse,
+  resolve,
+  sep,
+} = require('path');
+
+async function cpFn(src, dest, opts) {
+  // Warn about using preserveTimestamps on 32-bit node
+  if (opts.preserveTimestamps && process.arch === 'ia32') {
+    const warning = 'Using the preserveTimestamps option in 32-bit ' +
+      'node is not recommended';
+    process.emitWarning(warning, 'TimestampPrecisionWarning');
+  }
+  const stats = await checkPaths(src, dest, opts);
+  const { srcStat, destStat } = stats;
+  await checkParentPaths(src, srcStat, dest);
+  if (opts.filter) {
+    return handleFilter(checkParentDir, destStat, src, dest, opts);
+  }
+  return checkParentDir(destStat, src, dest, opts);
+}
+
+async function checkPaths(src, dest, opts) {
+  const { 0: srcStat, 1: destStat } = await getStats(src, dest, opts);
+  if (destStat) {
+    if (areIdentical(srcStat, destStat)) {
+      throw new ERR_FS_CP_TO_SUBDIRECTORY({
```

Comment:
It would probably make sense to have a separate error for cases where `src and dest cannot be the same` since there are variety of cases that src and dest could be the same but they are not necessarily a subdirectory of themselves.

Reply:
I opted to introduce the more generic `ERR_FS_CP_EINVAL`, which works for both scenarios. The error message itself still varies.